feat(ai-gateway): Align semantic similarity documentation with AI GW 2.0 (#5501)
Kong/developer.konghq.com: app/ai-gateway/semantic-similarity.md

Reviewer: ```suggestion
However, extremes at the far ends of the spectrum present significant drawbacks:
```

    icon: /assets/icons/redis.svg
---

Vector embeddings represent text as points in high-dimensional space, where the distance between vectors reflects semantic similarity. This enables semantic search—comparing meaning rather than exact words—powering LLM workflows like intelligent caching, retrieval, classification, and anomaly detection.

![Vector embeddings example](/assets/images/ai-gateway/vectors.svg)
> _**Figure 1:** A simplified representation of vector text embeddings in a three-dimensional space._

> For example, in the image, “king” and “emperor” are semantically more similar than “king” is to “otter”. Similarity is measured using techniques like cosine similarity or Euclidean distance, which quantify the relationship between vectors.

Reviewer: ```suggestion
For example, in the figure 1, “king” and “emperor” are semantically more similar than “king” is to “otter”. Similarity is measured using techniques like cosine similarity or Euclidean distance, which quantify the relationship between vectors.
```

1. **Semantic load balancing**: Route requests to upstream providers based on how semantically similar the prompt is to each provider's capabilities, using the `semantic` load balancing algorithm.
2. **Semantic Policies**: Attach Policies like AI Semantic Cache or AI Semantic Prompt Guard to add similarity-based caching, retrieval-augmented generation (RAG), and guardrails.

### Vector databases

To store and compare embeddings efficiently, {{site.ai_gateway}} semantic features rely on vector databases. These specialized datastores index high-dimensional embeddings and enable **fast similarity search** based on distance metrics like cosine similarity or Euclidean distance.

A Model Entity’s [semantic load balancer](/ai-gateway/entities/ai-model/#algorithms) stores vector representations of each target model’s semantic description at configuration time, and uses the vector database to compare incoming prompts against those stored vectors. Semantic policies use the same vector database to perform similarity searches at request time.

> The selected database stores the embeddings generated by the Model or Policies (either at config time or runtime), and determines the accuracy and performance of semantic operations.

Reviewer: ```suggestion
A Model Entity’s [semantic load balancer](/ai-gateway/entities/ai-model/#algorithms) stores vector representations of each target model’s semantic description at configuration time, and uses the vector database to compare incoming prompts against those stored vectors. 

Semantic policies also use vector databases to perform similarity searches at request time. The selected database stores the embeddings generated by the Model or Policies (either at config time or runtime), and determines the accuracy and performance of semantic operations.
```

make the seperation cleaner and avoid confusion over the interpretation of 'same database'

> {% include plugins/ai-vector-db.md name="semantic features" %}

Reviewer: Needs the new includes location

## Semantic similarity in {{site.ai_gateway}}

> In {{site.ai_gateway}} {% new_in 2.0.0 %}, semantic similarity enables intelligent request routing, caching, and content filtering based on meaning rather than exact matches. A [Model](/ai-gateway/entities/ai-model/) can leverage semantic similarity in two ways:

Reviewer: - Do we need the `{% new_in 2.0.0 %}` here, there's no new capability 
- I would reverse the ordering here so that 'semantic similarity' is explained by the concept of 'meaning'

```suggestion
In {{site.ai_gateway}} {% new_in 2.0.0 %}, you can perform intelligent request routing, caching, and content filtering based on meaning rather than exact matches by using semantic similarity queries. A [Model](/ai-gateway/entities/ai-model/) can leverage semantic similarity in two ways:
```

1. **Semantic load balancing**: Route requests to upstream providers based on how semantically similar the prompt is to each provider's capabilities, using the `semantic` load balancing algorithm.
2. **Semantic policies**: Attach policies like AI Semantic Cache or AI Semantic Prompt Guard to add similarity-based caching, retrieval-augmented generation (RAG), and guardrails.

### Vector databases

> To compare embeddings efficiently, {{site.ai_gateway}} semantic features rely on vector databases. These specialized data stores index high-dimensional embeddings and enable **fast similarity search** based on distance metrics like cosine similarity or Euclidean distance.

Reviewer: ```suggestion
To store and compare embeddings efficiently, {{site.ai_gateway}} semantic features rely on vector databases. These specialized datastores index high-dimensional embeddings and enable **fast similarity search** based on distance metrics like cosine similarity or Euclidean distance.
```

> When a Model’s semantic load balancer or an attached semantic policy needs to find semantically similar content—whether it’s a prompt, a target model description, or a document chunk—it sends a query to a vector database. The database returns the closest matches, allowing the Model or policy to make decisions like routing, caching, injecting, or blocking.

Reviewer: This is redundant, please remove

```suggestion
```

To compare embeddings efficiently, {{site.ai_gateway}} semantic features rely on vector databases. These specialized data stores index high-dimensional embeddings and enable **fast similarity search** based on distance metrics like cosine similarity or Euclidean distance.

When a plugin needs to find semantically similar content—whether it’s a past prompt, an upstream description, or a document chunk—it sends a query to a vector database. The database returns the closest matches, allowing the plugin to make decisions like caching, routing, injecting, or blocking.
When a Model’s semantic load balancer or an attached semantic policy needs to find semantically similar content—whether it’s a prompt, a target model description, or a document chunk—it sends a query to a vector database. The database returns the closest matches, allowing the Model or policy to make decisions like routing, caching, injecting, or blocking.

> {% include_cached /plugins/ai-vector-db.md name=page.name %}

Reviewer: Why was the include of supported vector DBs removed, this is important and often requested information

Author: It was my mistake, I restored it

> A Model’s [semantic load balancer](/ai-gateway/entities/ai-model/#algorithms) stores vector representations of each target model’s semantic description at configuration time, and uses the vector database to compare incoming prompts against those stored vectors. Semantic policies use the same vector database to perform similarity searches at request time.

Reviewer: To disambiguate from upstream target models

```suggestion
A Model Entity’s [semantic load balancer](/ai-gateway/entities/ai-model/#algorithms) stores vector representations of each target model’s semantic description at configuration time, and uses the vector database to compare incoming prompts against those stored vectors. Semantic policies use the same vector database to perform similarity searches at request time.
```

> The selected database stores the embeddings generated by the Model or policies (either at config time or runtime), and determines the accuracy and performance of semantic operations. {{site.ai_gateway}} supports Redis with Vector Similarity Search (VSS) and PostgreSQL with the pgvector extension.

Reviewer: You capitalise 'Model' here but not 'policies', if you think we should capitalise entities then that should be both.

Reviewer: Restore the supported vector DBs include or create a new one to be used here

Reviewer: Consider linking to Redis and Postgres docs

Author: All of these are fixed now.

> Semantic similarity is applied differently depending on the feature:

Reviewer: ```suggestion
Semantic similarity is used differently depending on the feature:
```

> **Model semantic load balancing** (`semantic` algorithm):

Reviewer: formatting of this heading looks quite strange

Author: Fixed

> - Compares incoming prompt embeddings against stored embeddings of each target model's semantic description.

Reviewer: Is this the complete description? Based on my knowledge the described process would not have great results.

Author: I've improved this one, Does it look okay now?

> - Each semantic policy applies similarity search slightly differently based on its goal.

Reviewer: ```suggestion
- Each semantic policy uses similarity search slightly differently based on its goal.
```

> - AI Semantic Prompt Guard and AI Semantic Response Guard compare content against allow and deny lists to detect misuse patterns.

Reviewer: Distinguish from regular prompt guard

```suggestion
- AI Semantic Prompt Guard and AI Semantic Response Guard compare content against vectorised allow and deny lists to detect misuse patterns semantically.
```

> Dimensionality determines how many numerical features represent each piece of content—similar to how a detailed profile might have dimensions for age, interests, location, and preferences. Higher dimensions create more detailed "fingerprints" that capture nuanced relationships, with smaller distances between vectors indicating stronger conceptual similarity and larger distances showing weaker associations.

Reviewer: ```suggestion
Dimensionality determines how many numerical features represent each piece of content—similar to how a detailed profile might have dimensions for age, interests, location, and preferences. A higher number of dimensions creates more detailed "fingerprints" that capture nuanced relationships. Smaller distances between vectors indicate stronger conceptual similarity and larger distances show weaker associations.
```
